### PR TITLE
Introduce and extend Config pattern (III/2)

### LINKED ISSUE #1780: Implement builder pattern of JedisPool and JedisCluster
### Expected behavior

Currently, the `JedisCluster` constructor has at most 7 parameter.
Now
```java
  public JedisCluster(HostAndPort node, int connectionTimeout, int soTimeout,
          int maxAttempts, String password, String clientName, final GenericObjectPoolConfig poolConfig) {
  }
```

After adding support SSL in `JedisCluster` #1550, the number become 12

```java
  public JedisCluster(Set<HostAndPort> jedisClusterNode, int connectionTimeout, int soTimeout,
                      int maxAttempts, String password, String clientName, final GenericObjectPoolConfig poolConfig,
                      boolean ssl, SSLSocketFactory sslSocketFactory, SSLParameters sslParameters, 
                      HostnameVerifier hostnameVerifier, JedisClusterHostAndPortMap hostAndPortMap) {
  }
```

We should implement builder pattern to make it easy to build.

Same as `JedisPool`

### LINKED ISSUE #1824: if uri contains dbIndex, timeout configuration doesn't work
### Expected behavior
set timeout to 10000 ms

### Actual behavior
timeout is 2000 ms as default

### Steps to reproduce:

1 .URI uri = new URI("redis://13.229.188.59:6379/1");//use fake address to test timeout
    Jedis jedis = new Jedis(uri, 10000);
2. when time passed 2 seconds, throws SocketTimeoutException, That's not my expected.
3. URI uri = new URI("redis://13.229.188.59:6379");//use fake address to test timeout
    Jedis jedis = new Jedis(uri, 10000);
    jedis.flushDB();
4. when time passed 10 seconds, throws SocketTimeoutException, That's expected.

### LINKED ISSUE #2234: Support host and port mapping for cluster connections (not only SSL)
### Expected behavior

When connecting to a cluster which is run inside Docker (e.g. using http://testcontainers.org library, which maps ports dynamically depending on available ones) `JedisCluster` should take a mapping function to properly map discovered nodes host/port to their real ports.

### Actual behavior

Jedis connects to the node(s) using host/port specified in `JedisCluster` constructor, then discovers all nodes and fails when trying to connect to them directly.

### Proposed solution

Use `JedisClusterHostAndPortMap` not only for SSL connections. Is it a way to go?

### PR DESCRIPTION
**Pull request by sazzad16**
Resolves #1780 
Resolves #1824 
Resolves #2234 

Closes #1787 
Closes #2237 

Closes #2342 
Closes #2349 
Closes #2352 
Closes #2353 
Closes #2363

### CODE REVIEW

File: src/main/java/redis/clients/jedis/DefaultJedisSocketFactory.java [hunk 1]
```diff
@@ -1,80 +1,122 @@
...
+  public DefaultJedisSocketFactory(HostAndPort hostAndPort, JedisClientConfig config) {
+    setHostAndPort(hostAndPort);
+    if (config != null) {
+      setConnectionTimeout(config.getConnectionTimeout());
+      setSoTimeout(config.getSoTimeout());
+      setSsl(config.isSsl());
+      setSslSocketFactory(config.getSslSocketFactory());
+      setSslParameters(config.getSslParameters());
+      setHostnameVerifier(config.getHostnameVerifier());
+      setHostAndPortMapper(config.getHostAndPortMapper());
```
Comment (mina-asham) [inline src/main/java/redis/clients/jedis/DefaultJedisSocketFactory.java:52]: It's a bit unconventional to use setters in constructors, any advantage of doing that?

File: src/test/java/redis/clients/jedis/tests/JedisClusterTest.java [hunk 2]
```diff
@@ -510,24 +566,35 @@ public void testCloseable() throws IOException {
...
+//    Iterator<JedisPool> poolIterator = jc.getClusterNodes().values().iterator();
+//    while (poolIterator.hasNext()) {
+//      JedisPool pool = poolIterator.next();
+//      try {
+//        pool.getResource();
+//        fail("JedisCluster's internal pools should be already destroyed");
+//      } catch (JedisConnectionException e) {
+//        // ok to go...
+//      }
+//    }
```
Comment (mina-asham) [inline src/test/java/redis/clients/jedis/tests/JedisClusterTest.java:586]: I think you forgot to uncomment this?

File: src/test/java/redis/clients/jedis/tests/JedisTest.java [hunk 3]
```diff
@@ -190,14 +179,40 @@ public void allowUrlWithNoDBAndNoPassword() {
...
+    URI uri = new URI("redis://fakehost:6378/1");
+    long startTime = System.nanoTime();
+    try (Jedis j = new Jedis(uri, 5000)) {
+      j.ping();
+    } catch (Exception ex) {
+      assertEquals(JedisConnectionException.class, ex.getClass());
+      assertEquals(java.net.UnknownHostException.class, ex.getCause().getClass());
+    }
+    long stopTime = System.nanoTime();
+    assertTrue(stopTime - startTime > 4000);
```
Comment (mina-asham) [inline src/test/java/redis/clients/jedis/tests/JedisTest.java:193]: I don't get this one, the timeout is 5 seconds, but we are validating the runtime is greater than 4 **milli**seconds?

File: src/test/java/redis/clients/jedis/tests/ShardedJedisPoolTest.java [hunk 4]
```diff
@@ -136,6 +137,7 @@ public void checkFailedJedisServer() {
     pool.destroy();
   }
 
+  @Ignore
```
Comment (mina-asham) [inline src/test/java/redis/clients/jedis/tests/ShardedJedisPoolTest.java:140]: Why ignore it? If the test doesn't make sense anymore can we remove it instead?

File: src/test/java/redis/clients/jedis/tests/ShardedJedisPoolWithCompleteCredentialsTest.java [hunk 5]
```diff
@@ -145,6 +146,7 @@ public void checkFailedJedisServer() {
     pool.destroy();
   }
 
+  @Ignore
```
Comment (mina-asham) [inline src/test/java/redis/clients/jedis/tests/ShardedJedisPoolWithCompleteCredentialsTest.java:149]: Ditto

**Review by mina-asham — CHANGES_REQUESTED** [review]
Looks good, thanks for going through the comments from the prevision revision! Just a few more minor comments here and then I am happy to sign off

File: src/main/java/redis/clients/jedis/DefaultJedisSocketFactory.java (hunk 1 above)
Comment (sazzad16) [reply src/main/java/redis/clients/jedis/DefaultJedisSocketFactory.java:52]: Not exactly. I saw this style is being used in commons-pool2. So I put it here at least to have some discussions :)

File: src/test/java/redis/clients/jedis/tests/JedisTest.java (hunk 3 above)
Comment (sazzad16) [reply src/test/java/redis/clients/jedis/tests/JedisTest.java:193]: Socket connections are dependent on OS. We can't strictly check the socket timeouts. I'm using the threshold to reduce this test's probability to become another [flaky](https://github.com/redis/jedis/issues/2367) test.

File: src/test/java/redis/clients/jedis/tests/ShardedJedisPoolTest.java (hunk 4 above)
Comment (sazzad16) [reply src/test/java/redis/clients/jedis/tests/ShardedJedisPoolTest.java:140]: I couldn't make this work with my current changes. What if someone else can? So I kept this.

File: src/test/java/redis/clients/jedis/tests/ShardedJedisPoolWithCompleteCredentialsTest.java (hunk 5 above)
Comment (sazzad16) [reply src/test/java/redis/clients/jedis/tests/ShardedJedisPoolWithCompleteCredentialsTest.java:149]: Ditto

File: src/test/java/redis/clients/jedis/tests/JedisClusterTest.java (hunk 2 above)
Comment (sazzad16) [reply src/test/java/redis/clients/jedis/tests/JedisClusterTest.java:586]: Nope.

I debugged and found that `jc.getClusterNodes().values()` is actually empty, even without my changes. So above codes are meaningless.

Also check that following `assertEquals(0, jc.getClusterNodes().size());` is passing.

File: src/main/java/redis/clients/jedis/DefaultJedisSocketFactory.java (hunk 1 above)
Comment (mina-asham) [reply src/main/java/redis/clients/jedis/DefaultJedisSocketFactory.java:52]: IMO, still unconventional commons-pool2 doesn't make it okay :P

File: src/test/java/redis/clients/jedis/tests/JedisClusterTest.java (hunk 2 above)
Comment (mina-asham) [reply src/test/java/redis/clients/jedis/tests/JedisClusterTest.java:586]: If that part is useless let's delete it then, commenting it out leave unanswered questions

File: src/test/java/redis/clients/jedis/tests/JedisTest.java (hunk 3 above)
Comment (mina-asham) [reply src/test/java/redis/clients/jedis/tests/JedisTest.java:193]: Got you, so that would make it "less" flaky, but potentially still flaky, do we want to avoid that check completely? We are kind of checking the JDK's Socket timeout works as expected here which is not really adding much, I think just checking the exception is enough in that case

File: src/test/java/redis/clients/jedis/tests/ShardedJedisPoolTest.java (hunk 4 above)
Comment (mina-asham) [reply src/test/java/redis/clients/jedis/tests/ShardedJedisPoolTest.java:140]: If it worked before your change then there are two cases:
1. We are testing the wrong thing, in that case let's drop the test
2. We broke something so w need to look into it

I think in either cases we need to look into this one and the other one below

File: src/test/java/redis/clients/jedis/tests/JedisTest.java (hunk 3 above)
Comment (sazzad16) [reply src/test/java/redis/clients/jedis/tests/JedisTest.java:193]: I really want to have test for #1824 as this is a reported issue. And I can't find any other way to test it.

Comment (mina-asham) [reply src/test/java/redis/clients/jedis/tests/JedisTest.java:193]: This should do the trick? (a bit more involved, but should be a lot more stable)

```suggestion
    int port = 6677;
    int timeoutMillis = 5000;
    int deltaMillis = 500;

    URI uri = new URI("redis://localhost:" + port + "/1");
    ServerSocket server = new ServerSocket(port);
    Instant start = Instant.now();
    try (Jedis jedis = new Jedis(uri, timeoutMillis)) {
        jedis.ping();
    } catch (JedisConnectionException ex) {
        assertEquals(SocketTimeoutException.class, ex.getCause().getClass());
        assertEquals(timeoutMillis, Duration.between(start, Instant.now()).toMillis(), deltaMillis);
        return;
    } finally {
        server.close();
    }
    fail();
```

Comment (sazzad16) [reply src/test/java/redis/clients/jedis/tests/JedisTest.java:193]: I'll give that a try.

File: src/main/java/redis/clients/jedis/DefaultJedisClientConfig.java [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+
+import javax.net.ssl.HostnameVerifier;
+import javax.net.ssl.SSLParameters;
+import javax.net.ssl.SSLSocketFactory;
+
+public final class DefaultJedisClientConfig implements JedisClientConfig {
+
+  private final int connectionTimeout;
+  private final int soTimeout;
+  private final int infiniteSoTimeout;
```
Comment (walles) [inline src/main/java/redis/clients/jedis/DefaultJedisClientConfig.java:11]: With Java 8 (#2357) `Duration` could be used rather than `int`s for the timeouts, which would probably be a better API in the long run.

Comment (walles) [reply src/main/java/redis/clients/jedis/DefaultJedisClientConfig.java:11]: I assume this goes for all places where `Protocol.DEFAULT_TIMEOUT` is used really, as it is now the unit for the timeouts is not obvious.

Comment (sazzad16) [reply src/main/java/redis/clients/jedis/DefaultJedisClientConfig.java:11]: @walles Let's merge this PR first and then we'll think about it :)

**Review by mina-asham — APPROVED** [review]

File: src/main/java/redis/clients/jedis/BinaryClient.java [hunk 7]
```diff
@@ -59,16 +65,28 @@ public BinaryClient(final String host, final int port) {
...
+  /**
+   * @deprecated This constructor will be removed in future.
+   */
+  @Deprecated
   public BinaryClient(final String host, final int port, final boolean ssl) {
     super(host, port, ssl);
   }
 
+  /**
+   * @deprecated This constructor will be removed in future.
```
Comment (gkorland) [inline src/main/java/redis/clients/jedis/BinaryClient.java:77]: @sazzad16 please point to the alternative Constructor

File: src/main/java/redis/clients/jedis/BinaryClient.java [hunk 8]
```diff
@@ -59,16 +65,28 @@ public BinaryClient(final String host, final int port) {
     super(host, port);
   }
 
+  /**
+   * @deprecated This constructor will be removed in future.
```
Comment (gkorland) [inline src/main/java/redis/clients/jedis/BinaryClient.java:69]: @sazzad16 please point to the alternative Constructor

File: src/main/java/redis/clients/jedis/BinaryClient.java [hunk 9]
```diff
@@ -81,23 +99,20 @@ public boolean isInWatch() {
...
-
-  private byte[][] joinParameters(byte[] first, byte[] second, byte[][] rest) {
-    byte[][] result = new byte[rest.length + 2][];
-    result[0] = first;
-    result[1] = second;
-    System.arraycopy(rest, 0, result, 2, rest.length);
-    return result;
+  /**
+   * @param user
+   * @deprecated This method will be removed in future.
```
Comment (gkorland) [inline src/main/java/redis/clients/jedis/BinaryClient.java:104]: @sazzad16 please point to the alternative Constructor

File: src/main/java/redis/clients/jedis/BinaryClient.java [hunk 10]
```diff
@@ -81,23 +99,20 @@ public boolean isInWatch() {
...
+  @Deprecated
+  public void setUser(final String user) {
+    this.user = user;
   }
 
-  public void setUser(final String user) { this.user = user; }
-
+  /**
+   * @param password
+   * @deprecated This method will be removed in future.
```
Comment (gkorland) [inline src/main/java/redis/clients/jedis/BinaryClient.java:113]: @sazzad16 please point to the alternative Constructor

File: src/test/java/redis/clients/jedis/tests/ShardedJedisPoolTest.java [hunk 11]
```diff
@@ -136,49 +137,6 @@ public void checkFailedJedisServer() {
     pool.destroy();
   }
 
-  @Test
-  public void shouldReturnActiveShardsWhenOneGoesOffline() {
```
Comment (gkorland) [inline src/test/java/redis/clients/jedis/tests/ShardedJedisPoolTest.java:140]: @sazzad16 why did your remove this test?

File: src/test/java/redis/clients/jedis/tests/ShardedJedisPoolWithCompleteCredentialsTest.java [hunk 12]
```diff
@@ -145,49 +146,6 @@ public void checkFailedJedisServer() {
     pool.destroy();
   }
 
-  @Test
-  public void shouldReturnActiveShardsWhenOneGoesOffline() {
```
Comment (gkorland) [inline src/test/java/redis/clients/jedis/tests/ShardedJedisPoolWithCompleteCredentialsTest.java:149]: @sazzad16 why did your remove this test?

File: src/test/java/redis/clients/jedis/tests/ConnectionTest.java [hunk 13]
```diff
@@ -4,51 +4,60 @@
...
 
-  @Before
-  public void setUp() throws Exception {
-    client = new Connection();
-  }
-
   @After
   public void tearDown() throws Exception {
-    client.close();
+    if (client != null) {
```
Comment (gkorland) [inline src/test/java/redis/clients/jedis/tests/ConnectionTest.java:20]: @sazzad16 why do you still need this `@After` if the connection is not created in the `@Before` any more?

File: src/main/java/redis/clients/jedis/JedisSocketFactory.java [hunk 14]
```diff
@@ -15,23 +16,29 @@
  */
 public interface JedisSocketFactory {
```
Comment (gkorland) [inline src/main/java/redis/clients/jedis/JedisSocketFactory.java:17]: @sazzad16 it seems like all the methods are deprecated so how come the interface isn't?

File: src/test/java/redis/clients/jedis/tests/ShardedJedisTest.java [hunk 15]
```diff
@@ -270,9 +270,9 @@ public void testMasterSlaveShardingConsistency() {
         Hashing.MURMUR_HASH);
 
     List<JedisShardInfo> otherShards = new ArrayList<JedisShardInfo>(3);
-    otherShards.add(new JedisShardInfo("otherhost", Protocol.DEFAULT_PORT));
-    otherShards.add(new JedisShardInfo("otherhost", Protocol.DEFAULT_PORT + 1));
-    otherShards.add(new JedisShardInfo("otherhost", Protocol.DEFAULT_PORT + 2));
+    otherShards.add(new JedisShardInfo("127.0.0.1", Protocol.DEFAULT_PORT));
```
Comment (gkorland) [inline src/test/java/redis/clients/jedis/tests/ShardedJedisTest.java:273]: why did you change all the host names to IPs?

File: src/main/java/redis/clients/jedis/BinaryClient.java (hunk 9 above)
Comment (sazzad16) [reply src/main/java/redis/clients/jedis/BinaryClient.java:104]: @gkorland There won't be any.

File: src/main/java/redis/clients/jedis/BinaryClient.java (hunk 10 above)
Comment (sazzad16) [reply src/main/java/redis/clients/jedis/BinaryClient.java:113]: @gkorland There won't be any direct alternate.

File: src/main/java/redis/clients/jedis/JedisSocketFactory.java (hunk 14 above)
Comment (sazzad16) [reply src/main/java/redis/clients/jedis/JedisSocketFactory.java:17]: @gkorland `createSocket()` method is here to stay. But instead of throwing IOException, it will throw only JedisConnectionException.

File: src/test/java/redis/clients/jedis/tests/ConnectionTest.java (hunk 13 above)
Comment (sazzad16) [reply src/test/java/redis/clients/jedis/tests/ConnectionTest.java:20]: @gkorland This test class mostly consists of constructor tests. So connections are supposed to be created in each test methods. I'm closing those objects here in `@After`.

File: src/main/java/redis/clients/jedis/BinaryClient.java (hunk 9 above)
Comment (sazzad16) [reply src/main/java/redis/clients/jedis/BinaryClient.java:106]: @gkorland There won't be any direct alternate. Other than config class and, may be, factory class, no other class will hold user data.

File: src/main/java/redis/clients/jedis/BinaryClient.java (hunk 10 above)
Comment (sazzad16) [reply src/main/java/redis/clients/jedis/BinaryClient.java:115]: @gkorland There won't be any direct alternate. Other than config class and, may be, factory class, no other class will hold user data.

**Comment by sazzad16** [comment]
@gkorland 

1. ShardedJedisPoolTest.shouldReturnActiveShardsWhenOneGoesOffline()
2. ShardedJedisPoolWithCompleteCredentialsTest.shouldReturnActiveShardsWhenOneGoesOffline()
3. ShardedJedisTest.testMasterSlaveShardingConsistency()
4. ShardedJedisTest.testMasterSlaveShardingConsistencyWithShardNaming()

These tests used to work because underlying connections were not always created given the combination of parameters. All the Jedis classes were suffering from this situation. The underlying connections of objects representing **a connection** either should always be created from constructors or should NEVER be created. In this PR I implemented the former approach.

I was able to keep the tests with a valid host address in (3) and (4).

But I could not make (1) and (2) to work. Even though I `@Ignore`d first, I removed those after review. A portion of (1) and (2):
```
    shards.set(1, new JedisShardInfo("localhost", 1234));
    pool = new ShardedJedisPool(redisConfig, shards);
    jedis = pool.getResource();
```
This is more logical that `pool.getResource()` would fail because `1234` is an invalid port. But the tests are expecting `pool.getResource()` without any error. That's why these two tests won't work with this PR.

IMHO, we shouldn't care about impractical tests. Still, I can `@Ignore` those if you prefer.

File: src/test/java/redis/clients/jedis/tests/ShardedJedisPoolTest.java (hunk 11 above)
Comment (sazzad16) [reply src/test/java/redis/clients/jedis/tests/ShardedJedisPoolTest.java:140]: @gkorland https://github.com/redis/jedis/pull/2368#issuecomment-790262400

File: src/test/java/redis/clients/jedis/tests/ShardedJedisPoolWithCompleteCredentialsTest.java (hunk 12 above)
Comment (sazzad16) [reply src/test/java/redis/clients/jedis/tests/ShardedJedisPoolWithCompleteCredentialsTest.java:149]: @gkorland https://github.com/redis/jedis/pull/2368#issuecomment-790262400

File: src/test/java/redis/clients/jedis/tests/ShardedJedisTest.java (hunk 15 above)
Comment (sazzad16) [reply src/test/java/redis/clients/jedis/tests/ShardedJedisTest.java:273]: @gkorland https://github.com/redis/jedis/pull/2368#issuecomment-790262400

File: src/main/java/redis/clients/jedis/BinaryClient.java (hunk 9 above)
Comment (gkorland) [reply src/main/java/redis/clients/jedis/BinaryClient.java:106]: So you should state in the JDoc

File: src/main/java/redis/clients/jedis/BinaryClient.java (hunk 10 above)
Comment (gkorland) [reply src/main/java/redis/clients/jedis/BinaryClient.java:115]: So you should state it in the JDoc

File: src/main/java/redis/clients/jedis/JedisSocketFactory.java (hunk 14 above)
Comment (gkorland) [reply src/main/java/redis/clients/jedis/JedisSocketFactory.java:17]: Yes, since you are not going to really deprecate it

**Review by gkorland — APPROVED** [review]
